### DISCUSSION TITLE
Include hardware critical options in templates

### DISCUSSION THREAD

**Discussion by Teslafly** [arendst/Tasmota#11788] [Ideas]
There are some templates that require an option to go with them for the hardware to work correctly.

For example, many bulbs/led drivers require Option92 (PWMCT), SetOption37 (ChannelRemap), or  Option68 (MultiPwm) to be set to work correctly.
My intuition is that a template should include everything needed for the hardware to work correctly.

It would be nice if it was possible to attach the setting of a non default option to a config
possibly something like:

base:
`{"NAME":"OM100/RGBWW","GPIO":[0,0,0,0,416,419,0,0,417,452,418,0,0,0],"FLAG":0,"BASE":18}`

with options:
`{"NAME":"OM100/RGBWW","GPIO":[0,0,0,0,416,419,0,0,417,452,418,0,0,0],"FLAG":0,"BASE":18, OPTION[92:1,37:5]}`
would additionally set option 92 to 1, and 37 to 5

Is this something worth considering? is it already possible and the output template doesn't include these hardware critical option settings?

**Comment by arendst** [comment]
It's worth considering for an input template. To make it work for an output template would be a challenge how to find which of over 100 setoptions need to be specifically set for this template...

Perhaps someone comes up with a very clever, low memory usage solution.

**Reply by sfromis** [reply]
Some relevant setup options like debouncing or PWM setup are not `SetOption` numbers. One low memory method could be for a runtime template copy/paste to include a Backlog-style command string, executed right away, not saved, without the possibility of capturing for automatically merging a new template.

While this could even be handled transparently in Tasmotizer, not having the copy/paste may be going a bit far.

**Reply by barbudor** [reply]
Would it be possible to find a way to easily embedded autoexec.bat (or a firststart.bat) into the flashing process ?
I've not looked at how the bin file is structured but may be it would be possible to have a very strict minimal file system embedded ?
Still remains the problem of devices with only 1MB as we said this is too low for a filesystem.
May be not a full featured filesystem but a read-only thing with only 1 file limited to let's ay 4KB ?

**Reply by sfromis** [reply]
You'd need no file system shim if able to embed a backlog style command sequence. Probably primarily for copy/paste with a precompiled binary for "common user easy config". With a file system, the default command could be 
`UfsRun autoexec.bat`

**Reply by sfromis** [reply]
Another use case for having commands available in templates could be for an "inclusive" build avoiding sensor and driver conflicts by having `I2CDRIVERS_0_31` (and the others) all zero, with suitable `I2CDriver` enabling commands in a copy/paste template. Of course, these command could also be put into an initial scripted config setup or autoexec.bat, but the user-friendliness of a single copy/paste for device config is "another level". Personally, I'm doing the "inclusive" setup for my own needs, as this fits nicely with my existing configuration management strategy.

**Reply by arendst** [reply]
I think the first suggestion about adding a CMND json fieldpair with backlog commands to be used only at initial template load AND activate time is the best code saving option.

I 'll work on it. I also see difficulties in cases where a template is loaded but not activated at the same time as these commands will not be stored and therefore won't be executed at delayed activation time.

Let's see how useful this is as it is more functionality than we have now.

**Reply by sfromis** [reply]
Not sure how this fits the internal structure, but ideally the command string could also be included in the build time template string `USER_TEMPLATE` to allow "factory reset" to revert to full initial config for some device, without needing extra steps. Of course, this could imply a quite long string also including setting up required rules. The latter may be a bit tricky due to semicolon also being required for rules, not fitting entirely into a backlog-style command string. As long as a template command string would not have to be stored in settings, a lot can be done....

**Reply by arendst** [reply]
I'll work on this:
```
{"NAME":"OM100/RGBWW","GPIO":[0,0,0,0,416,419,0,0,417,452,418,0,0,0],"FLAG":0,"BASE":18,"CMND":"SO91 1;SO37 5"}
```
and if the CMND part is not too long it may even be stored in the character pool as long as there is space available. For instance my pool is normally just half filled leaving a whopping 300 characters for a small CMND.

No need to add the backlog command as that will be added by the code.

In theory the additional CMND part could be entered by the template define too as currently this template define is executed byt he same JsonTemplate() function which needs to serve a "normal" template decode.

Using this new template on the current Tasmota version succesfully processes the template as an old type template:
```template
10:40:52.887 CMD: template
10:40:52.889 SRC: Serial
10:40:52.890 CMD: Group 0, Index 1, Command "TEMPLATE", Data ""
10:40:52.898 MQT: stat/wemos10/RESULT = {"NAME":"Generic","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}
template {"NAME":"OM100/RGBWW","GPIO":[0,0,0,0,416,419,0,0,417,452,418,0,0,0],"FLAG":0,"BASE":18,"CMND":"SO91 1;SO37 5"}
10:40:59.384 CMD: template {"NAME":"OM100/RGBWW","GPIO":[0,0,0,0,416,419,0,0,417,452,418,0,0,0],"FLAG":0,"BASE":18,"CMND":"SO91 1;SO37 5"}
10:40:59.388 SRC: Serial
10:40:59.389 CMD: Group 0, Index 1, Command "TEMPLATE", Data "{"NAME":"OM100/RGBWW","GPIO":[0,0,0,0,416,419,0,0,417,452,418,0,0,0],"FLAG":0,"BASE":18,"CMND":"SO91 1;SO37 5"}"
10:40:59.405 CFG: CR 337/699, Busy 0
10:40:59.414 MQT: stat/wemos10/RESULT = {"NAME":"OM100/RGBWW","GPIO":[0,0,0,0,416,419,0,0,417,452,418,0,0,0],"FLAG":0,"BASE":18}
10:40:59.844 CFG: Saved to flash at FB, Count 1466, Bytes 4096
```

In any way the CMND string needs to be as short as possible. I expect up to 30 chars. Do not want rules defined here. That would need to be a totally different way of configuration only possible with a filesystem simply because ther is not enough flash/RAM (rememeber all settings are realtime handeld in a RAM page currently 4k in size).

**Reply by sfromis** [reply]
Yeah, the restrictions coming from storing the template JSON in settings was the reason why I initially focused on processing the command string during copy/paste (and from compiled string), without the nicety of also saving the full string pasted into the web UI. Apart from the ability to have a "dormant" template stored, I'm not sure that I see that much value in having the string saved, but I may be missing the point here....

**Reply by meingraham** [reply]
After initial activation if the template and the initialization CMND is invoked, would the CMND string itself be cleared out of the character pool?

**Reply by barbudor** [reply]
If the template is hardcoded in the binary with `#define USER_TEMPLATE`, I would expect the CMND sequence to be re-executed after a reset 1/2.
If the template is entered in the GUI or console, it doesn't need to be saved as it wouldn't be re-executed any more.

**Reply by sfromis** [reply]
Yeah, with no "reset to initial + template only" command, the main utility of saving it would to aid the memory of the user, including reuse on another device.

**Reply by sfromis** [reply]
@meingraham Yes, the CR value in `Status 2` very much suggests that the CMND string is not included there, and the `Template` command displays the template without the original CMND included. The command seems to be completely gone after having the initial desired effect.

**Reply by arendst** [reply]
Correct. Currently the string is gone once executed.

Another issue is that versions before 9.4.0.2 will report a json error on template load using the gui because the CMND field contained a semicolon which was handled by backlog command. I'm reluctant to replace the semicolon in the CMND field by a "|" as that introduces another special case.

**Comment by arendst** [comment]
The latest dev release (9.4.0.2) adds support for basic template related commands by extending the template with the ``CMND`` fieldpair.

In case of the OP it would look like this:
```
{"NAME":"OM100/RGBWW","GPIO":[0,0,0,0,416,419,0,0,417,452,418,0,0,0],"FLAG":0,"BASE":18,"CMND":"SO91 1;SO37 5"}
```

It will only be executed if the template is selected as module 0, either before the template is loaded, when the activate box is selected OR if the ``module 0`` command is included in the ``CMND`` string.

**Reply by sfromis** [reply]
Just tested a template with CMND to establish a boot rule to yank ADPS9960 out of gesture mode. Worked nicely :-)

**Reply by digiblur** [reply]
That's down right awesome....

**Reply by nicandris** [reply]
tested also 👌  this simplifies things a lot in my setup! thanks for this, love it!!

**Comment by blakadder** [comment]
This is basically a backlog inside the template?

**Reply by arendst** [reply]
Yes but as the internal char array is very small only a few template related commands are supported.

Also, to provide backward compatibility the commands now need to be separated by a "|" character instead of the normal ";" character which would interfere with tasmota versions before 9.4.0.2 GUI template processing resulting in a JSON error.

**Reply by blakadder** [reply]
And how cmall is the array aka how many total chars can we fit in?

**Reply by arendst** [reply]
It all depends on how you enter the template and how many gpios have a different value than 0 or 1. The current tasmota input implemntations (serial, mqtt, gui other menu option, gui console and http al have different buffer sizes) define max total template length. This lenght is the complete template JSON message size.

- Using a serial console the max charactercount is dependant on the global serial input buffer being 520 characters including the ``template`` command.
- MQTT input is dependant on the max MQTT payload size being around 1040 characters, again in a global buffer.
- Gui other menu uses a global buffer fixed to 1200 characters.
- Gui console and HTTP use a global buffer fixed to 1200 characters max

All these max buffer sizes are the result of tuning tasmota memory, stack and heap usage and changing them will probably result in out-of-stack or out-of-heap exceptions.

So that's why I insist on using only a small number of template related commands. A complete rule of 500 chars (uncompressed) or even 800 chars compressed is a no go.

EDIT: Change code to support more characters in Gui other menu.
